### PR TITLE
Improve EI operator & add HPA for EI deployment.

### PR DESCRIPTION
**Pull request by abeykoon**
## Purpose

Improve merged EI operator to read parameters from CR files and if not given, refer to defaults specified at the configMap. 
Add Horizontal Auto Scaling support for EI controller. 

Now a sample CR for integration will look like below. 

```
apiVersion: wso2.com/v1alpha1
kind: Integration
metadata:
  name: test-integration
spec:
  replicas: 1
  image: abc/k8s-samples:helloservice
  deploySpec:
    minReplicas: 1
    requestCPU: "500m"
    reqMemory: "512Mi"
    cpuLimit: "2000m"
    memoryLimit: "2048Mi"
  autoScale:
    enabled: "true"
    maxReplicas: 3
  env:
    - name: DEV_ENV_USER_EP
      value: "https://reqres.in/api"
```

Sample config map definition for integration

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: ei-operator-config
data:
  host: wso2ei
  autoIngressCreation: "true"
  enableAutoScale: "false"
  minReplicas: "1"
  maxReplicas: "5"
  hpaMetrics: |
    - type: Resource
      resource:
        name: cpu
        target:
          type: Utilization
          averageUtilization: 50
```

### CODE REVIEW

File: api-operator/deploy/controller-configs/integration_controller_conf.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: ei-operator-config
```
Comment (sajithaliyanage) [inline api-operator/deploy/controller-configs/integration_controller_conf.yaml:20]: Shall we change the name as "integration-config"?

File: api-operator/deploy/controller-configs/integration_controller_conf.yaml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  enableAutoScale: "false"
+  minReplicas: "1"
+  maxReplicas: "5"
+  hpaMetrics: |
+    - type: Resource
+      resource:
+        name: cpu
+        target:
+          type: Utilization
+          averageUtilization: 50
```
Comment (sajithaliyanage) [inline api-operator/deploy/controller-configs/integration_controller_conf.yaml:33]: Let's add a new line at the end.

File: api-operator/pkg/controller/integration/integration_configmap.go [hunk 3]
```diff
@@ -19,32 +19,40 @@
...
-type EIController struct {
+// EIConfig control all deployments of the ei-ingress
+type EIConfig struct {
 	AutoCreateIngress bool
 	SSLRedirect       string
 	TLS               string
 	Host              string
+	EnableAutoScale   bool
+	MinReplicas       int32
+	MaxReplicas		  int32
```
Comment (sajithaliyanage) [inline api-operator/pkg/controller/integration/integration_configmap.go:36]: Fix the formatting issue

File: api-operator/pkg/controller/integration/integration_configmap.go [hunk 4]
```diff
@@ -68,6 +76,33 @@ func (r *ReconcileIntegration) UpdateDefaultConfigs(integration *wso2v1alpha1.In
 		if configMap.Data["ingressTLS"] != "" {
 			eic.TLS = configMap.Data["ingressTLS"]
 		}
+
+		if configMap.Data[enableAutoScaleKey] != "" {
+			eic.EnableAutoScale, err = strconv.ParseBool(configMap.Data[enableAutoScaleKey])
+			if err != nil {
+				log.Error(err,"Cannot parse enableAutoScaleKey to a boolean value. Setting false")
```
Comment (sajithaliyanage) [inline api-operator/pkg/controller/integration/integration_configmap.go:83]: Fix the formatting issue near the comma

File: api-operator/pkg/controller/integration/integration_configmap.go [hunk 5]
```diff
@@ -68,6 +76,33 @@ func (r *ReconcileIntegration) UpdateDefaultConfigs(integration *wso2v1alpha1.In
...
+				eic.EnableAutoScale = false
+			}
+		}
+
+		if configMap.Data[minReplicasKey] !="" {
```
Comment (sajithaliyanage) [inline api-operator/pkg/controller/integration/integration_configmap.go:88]: Fix the formatting issue near !=

File: api-operator/deploy/controller-configs/integration_controller_conf.yaml (hunk 1 above)
Comment (abeykoon) [reply api-operator/deploy/controller-configs/integration_controller_conf.yaml:20]: +1. Updated.

File: api-operator/deploy/controller-configs/integration_controller_conf.yaml (hunk 2 above)
Comment (abeykoon) [reply api-operator/deploy/controller-configs/integration_controller_conf.yaml:33]: Done

File: api-operator/pkg/controller/integration/integration_configmap.go (hunk 3 above)
Comment (abeykoon) [reply api-operator/pkg/controller/integration/integration_configmap.go:36]: Fixed.

File: api-operator/pkg/controller/integration/integration_configmap.go (hunk 4 above)
Comment (abeykoon) [reply api-operator/pkg/controller/integration/integration_configmap.go:83]: Updated.

File: api-operator/pkg/controller/integration/integration_configmap.go (hunk 5 above)
Comment (abeykoon) [reply api-operator/pkg/controller/integration/integration_configmap.go:88]: Updated.

**Review by sajithaliyanage — APPROVED** [review]

**Pull request merged**